File: RESTful Engine/Client Code/v2 API/JavaScript-RESTful-Client-Sample/generate-document.js
import pkg from 'windwardrestapi';
import fs from "fs";
const {WindwardClient, Template, Xml_10DataSource, OutputFormatEnum} = pkg;

const config = JSON.parse(fs.readFileSync('./config.json'));

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
};

async function main()
{
    /**
     * Step 1: INITIALIZE THE RESTFUL CLIENT AND SET UP THE TEMPLATE AND DATA SOURCE
     * --------------------------------------------------------------------------------------------
     */
    console.log("Initializing Fluent RESTful Client...");

    // Create a new instance of the client using the RESTful Engine URL provided in the config file
    let client = new WindwardClient.WindwardClient(config['restful-engine-url']);

    // Pass the license to the client.  The license should be specified in the config file
    // If you don't have a license key, you can leave it blank and it will produce output with a watermark.
    client.license = config['license'];
    
    // Display the version info for the restful engine
    let version = await client.getVersionInfo();
    console.log("Fluent RESTFul Engine Version: ", version);

    // The datasource document I am sending to the engine to be processed
    const dataSourcePath = './files/InvestmentFactSheet.xml';

    // Create the object and pass in the datasource name, and the data file path.
    let dataSource = new Xml_10DataSource('InvestmentFactSheet', undefined, dataSourcePath, undefined);
 
    //The template file I wish to process.
    const templatePath = './files/InvestmentFactSheet.docx';

    //Create a new template object (to see all the input params check the documentation)
    let template = new Template(OutputFormatEnum.PDF, [dataSource], undefined, templatePath,
        undefined, undefined, undefined, undefined, undefined, undefined,
        undefined, undefined, undefined, undefined, undefined, undefined);

    /**
     * Step 2: GENERATE THE DOCUMENT
     * --------------------------------------------------------------------------------------------
     */
    console.log("Generating Document...");

    // Post document to the engine for processing
    let document = await client.postDocument(template);
    
    //check postDocument status and wait if not ready.
    while(true) {
        await sleep(1000);
        let status = await client.getDocumentStatus(document.Guid);
        if (status == 302) {
            // The document generation is complete, we can now proceed to retrieve the generated document
            break;
        }
        else if (status == 201 || status == 202 || status == 404) {
            // The document generation is still in progress, continue waiting
            break;
        }
        else {
            // Potentially have an error, proceed to retrieve the document to get error details
            console.error("Error retrieving document. Status code: ", status);
            break;
        }
    }

    /**
     * Step 3: RETRIEVE THE DOCUMENT
     * --------------------------------------------------------------------------------------------
     */
    console.log("Retrieving Document...");

    // Retrieve the generated document
    let generatedDocument = await client.getDocument(document.Guid);

    // Write the processed document to a file (ensure the file format matches the output format specified in the template)
    fs.writeFile("./files/output.pdf", new Buffer.from(generatedDocument.Data, "base64"), function(err){});
    console.log("Generated document saved to /files/output.pdf");

    // Delete the processed document from the engine
    await client.deleteDocument(generatedDocument.Guid);
}

// Execute the main() method
main();

File: RESTful Engine/Client Code/v2 API/RESTful-Client-Python-Sample/src/fluent_demo/GenerateDocument.py
from .config import load_config, project_root, require_file
from windwardrestapi.Model import Template, Xml_10DataSource
from windwardrestapi.Api import WindwardClient as client
from pathlib import Path
from requests import HTTPError
import base64
import zipfile
import time


def main():
    #Read in RESTful server URL and license key from Config file
    cfg = load_config()

    fluent = cfg["fluent"]
    paths = cfg["paths"]

    base_url = fluent["base_url"]
    license_key = fluent["license_key"]

    template_path = require_file(project_root() / paths["template_file"], "template file")
    data_path = require_file(project_root() / paths["data_file"], "XML data file")

    # -------------- DOCUMENT WORK FLOW -------------- #

    # Initialize client and set license
    fluent_client = client.WindwardClient(base_url)
    fluent_client.licenseKey = license_key

    print(fluent_client.getVersion().toString())

    # Build the request body
    xml_ds = Xml_10DataSource.Xml_10DataSource(name="MANF_DATA_2009", data=data_path)
    template = Template.Template(
        data=template_path,
        outputFormat=Template.outputFormatEnum.DOCX,
        datasources=[xml_ds],
    )

    # POST, check status for completion
    try:
        doc = fluent_client.postDocument(template)
    except HTTPError as e:
        resp = getattr(e, "response", None)
        print("POST /v2/document failed")
        print("Status:", getattr(resp, "status_code", None))
        print("Body:", resp.text if resp is not None else "(no body)")
        raise

    print("GUID:", doc.guid)


    #OPTIONAL: Add timeout counter incase something goes wrong
    MAX_WAIT_SECONDS = 60
    start = time.time()
    deadline = start + MAX_WAIT_SECONDS
    #wait for document to generate, and check status for completion or error
    while True:
        #If we are passed the alloted time, end execution
        if time.time() >= deadline:
            print("Timeout waiting for document to be ready.")
            break

        # Sleep like the JS example: await sleep(1000)
        remaining = max(0, deadline - time.time())
        time.sleep(min(1, remaining))

        status = fluent_client.getDocumentStatus(doc.guid)

        if status == 302:
            # The document generation is complete, we can now proceed to retrieve the generated document
            print("Ready:", status)
            break
        elif status in (201, 202, 404):
            # The document generation is still in progress, continue waiting
            continue
        else:
            # Potentially have an error
            raise Exception("Error generating document. Status code:", status)
            break

    # Save output file to output directory
    out_dir = project_root() / "output"
    out_dir.mkdir(parents=True, exist_ok=True)
    file_path = out_dir / "output.docx"

    raw_bytes = fluent_client.getDocumentFile(doc.guid)
    with open(file_path, "wb") as fh:
        fh.write(raw_bytes)

    assert zipfile.is_zipfile(file_path), f"Not a valid DOCX/ZIP: {file_path}"
    with zipfile.ZipFile(file_path) as zf:
        assert zf.read("word/document.xml") is not None


if __name__ == "__main__":
    main()
